core/src/test/java/org/infinispan/api/APINonTxTest.java

```diff
@@ -661,4 +677,99 @@ public boolean equals(Object obj) {
       }
    }
 
+   void assertLockStream(BiConsumer<Cache<Object, Object>, CacheEntry<Object, Object>> consumer) throws Throwable {
+      for (int i = 0; i < 10; i++) {
+         cache.put(i, "value" + i);
+      }
+
+      CyclicBarrier barrier = new CyclicBarrier(2);
+
+      int key = 4;
+
+      LockedStream<Object, Object> stream = cache.getAdvancedCache().lockedStream();
+      Future<?> forEachFuture = fork(() -> stream.forEach((c, e) -> {
+         if (e.getKey().equals(key)) {
+            try {
+               barrier.await(10, TimeUnit.SECONDS);
```

Comment:
No they always run in the same order.

The `forEach` holds the key of the lock the entire time "outer" consumer duration including all calls to barrier.await.

The order goes like (a and b mean they can run at same time)

1. T1 submits stream
2a. T1 waits on barrier
2b. T2 acquires lock and waits on barrier
3a. T2 can run inner consumer and then waits on barrier while still holding the lock
3b. T3 can run put, but will be blocked
4. T1 checks timeout of T3 and finally release T2 and then waits until T2 completes
5. T2 completes consumer and finally releases lock

Maybe the confusion is the introduction of the second "inner" consumer?